### DISCUSSION TITLE
@Scheduled vs. Quartz JobExecutionContext

### DISCUSSION THREAD

**Discussion by famod** [quarkusio/quarkus#25987] [Q&A]
Hi @mkouba (and others),

as I understand from https://quarkusio.zulipchat.com/#narrow/stream/187030-users/topic/.40Scheduled.20annotation.20for.20Quartz.20Jobs/near/248450017, I have to use the programmatic API to `JobExecutionContext`.
Is that still the case?

What about a simple addition to `ScheduledExcution`, something like:
```java
Optional<T> getContext(Class<T> contextClass);
```
For Quartz one could then call `exec.getContext(JobExecutionContext.class).orElseThrow().[...]`.

Alternatively, there could be a simple, non-`Optional` getter only in `QuartzScheduledExecution`?

Our use case is storing some simple data with the job's `JobDataMap` and we really don't want to resort to the much more verbose programmatic approach...

Thanks!

[a comment by an automated account is not shown]

**Comment by mkouba** [comment]
Yes, we could make the `QuartzScheduledExecution` public and then add something like `public <T> T unwrap()`. So that you could do `scheduledExecution.<QuartzScheduledExecution>unwrap().context.getMergedJobDataMap()`...

@famod Could you pls file an issue?

**Reply by famod** [reply]
> Could you pls file an issue?

Sure, will do.

What puzzles me though (by merely looking at the code): The quartz job created via `@Scheduled` ist not stateful, right?
Stateful in the sense of `StatefulJob` that actually persists the data map?
So for my use case more changes are needed, no?

**Reply by mkouba** [reply]
Hm, so you need to persist the state between `@Scheduled` executions as if the job class would be annotated with `@PersistJobDataAfterExecution`, right?

**Reply by mkouba** [reply]
CC @machi1990

**Reply by machi1990** [reply]
I do not think the extension creates StatefulJob. That will have to be part of the change I suppose.

**Reply by mkouba** [reply]
[It's not](https://github.com/quarkusio/quarkus/blob/main/extensions/quartz/runtime/src/main/java/io/quarkus/quartz/runtime/QuartzScheduler.java#L563) and I don't think it should be the default...

**Reply by machi1990** [reply]
+1 something to be activated by a configuration? or to activate it all the time when the job store is a db one?

**Reply by famod** [reply]
I don't think it should depend on the store. Why shouldn't someone with the RAM store want to have the map stored between job executions? There is no real persistence due to the nature of the store, but still...

Ideally, one could control some details of the job config (like `@PersistJobDataAfterExecution` ) via an optional annotation, e.g. `@QuartzJobConfig(persistJobDataAfterExecution = true, ...)`, next to `@Scheduled`.
Alternatively or additionally, there could be a global config flag. Depends on how efficient storing the map is for jobs that don't need it...

**Reply by famod** [reply]
> Ideally, one could control some details of the job config (like `@PersistJobDataAfterExecution` ) via an optional annotation, e.g. `@QuartzJobConfig(persistJobDataAfterExecution = true, ...)`, next to `@Scheduled`.

Or something like `@QuartzJobClass(MyInvokerJob.class)` and the custom class has to extend `InvokerJob`. In this case the user can add the annotation on his own.

**Reply by famod** [reply]
#26062
